Rewrite termination measures sooner

bauereiss opened the pull request:
This used to happen after parsing, but now happens later in a rewrite. However, if it happens after other rewrites that rename variables, the measures might no longer type-check, so perform the measure rewrite as soon as possible.

bauereiss committed:
Rewrite termination measures sooner

This used to happen after parsing, but now happens later in a rewrite.
However, if it happens after other rewrites that rename variables, the
measures might no longer type-check, so perform the measure rewrite as
soon as possible.

bacam committed:
Correct rewrites about mutually recursive functions

Also adds a simple function call graph that is only function calls,
doesn't include any other dependencies, and doesn't need a typing.

bacam commented:
Two issues came up:
* The rewrite pass for minimising recursive function annotations wasn't handling mutually recursive definitions, which I've got a fix for.
* It separates this rewrite from the definition sorting, so if you use define measure function then it will go out of scope.  I'll try moving the valspec for any function directly used in the measure forward.  The measures I defined for the RISC-V are affected by this.

bacam committed:
Move valspecs for functions used in measures along with the measures

To ensure that the result will type check, we need to move any valspecs for functions
directly used in the measure forward.  The definitions themselves will be rearranged
later by the sorting rewrite.  Note that the type checker will ensure that a valspec
always exists.

bacam committed:
Add extra test for moving termination measures
